### DISCUSSION TITLE
Possible to register components after `Alpine.start()`?

### DISCUSSION THREAD

**Discussion by nikitaourazbaev** [alpinejs/alpine#3978] [1. Help] [answered]
I’m currently exploring using Alpine with [@11ty/is-land](https://github.com/11ty/is-land) but have run into needing this in the past as well, with components fetched from the server. At the time I couldn’t figure out a way to do it and ended up registering the components on the page that fetches the components’ HTML at the same time as the page’s own components.

Is there a way to register a component with `Alpine.data()` on a page after `Alpine.start()` has already happened, after `alpine:init` has been dispatched, and when other Alpine components are already running on the same page?

**Comment by ekwoka** [comment, marked as answer]
You can.

You just run `Alpine.data`. It can be used to add components after start with no issue.

What you will need to do is run that BEFORE adding the elements that depend on it to the page.

**Reply by nikitaourazbaev** [reply]
Ah, amazing! Not sure how I didn’t try that.

Thank you.

**Reply by nikitaourazbaev** [reply]
@ekwoka So this works really well for inserted HTML. Is something like this possible for HTML on the page?

Let’s say the element is already on the page at Alpine start but I want Alpine to ignore it until its Alpine component has been added with `Alpine.data()`. 

I’ve tried:

1. Not giving the element an `x-data` attribute and adding it programmatically after the component’s JS is loaded
2. Giving the element an `x-ignore` attribute and removing it programmatically after the component’s JS is loaded

Neither of those seem to work. But if I clone the element first, modify the attributes of the clone and replace the original element with the clone, that works. Is there a more elegant approach to get Alpine to discover the element?

_Update_:
Looks like `Alpine.initTree(deferredEl)` on the next tick works, is calling `initTree` directly a bad idea? 

Basically doing this:
```js
const deferredEl = document.querySelector('[x-data="TestComponent"]');
deferredEl.removeAttribute('x-ignore');

Alpine.nextTick(() => {
  Alpine.initTree(deferredEl);
});
```

**Reply by ekwoka** [reply]
Why would stuff already on the page not have Alpine.data already loaded?

How did you end up in such a situation where start was called first?

**Reply by nikitaourazbaev** [reply]
I’m experimenting with islands / partial hydration, where the HTML for all components is already on the page but the JS for some is deferred until they’re visible or interacted with, while others are registered and initialized right away (which is why start would already have been called).

In this case the Alpine.data for these deferred components would be in a JS file that’s loaded on demand.

(The context is a Shopify theme which can have an unknown but potentially large amount of sections on a page, some of which are Alpine components. I’d like to defer loading the JS for sections that are intensive or that are lower down on the page while having things like the header and first few sections be available and interactive right away.)

**Reply by ekwoka** [reply]
Okay.

You can have the branch have x-ignore and around it using something like x-intersect to load the component, add the data and evaluate the element.

**Reply by nikitaourazbaev** [reply]
The “evaluate the element” part is the one I’m a bit stuck on, everything up to that point I’ve been able to do. Removing `x-ignore` or adding `x-data` is not enough to trigger evaluation, seems like the element needs to either be removed and reinserted into the DOM, triggering a mutation, or I’ve discovered that I can run `initTree` directly on the element one tick after removing `x-ignore`. 

I’m wondering if running `initTree` directly is a bad idea or if there’s another way to get Alpine to evaluate the element after removing `x-ignore`.

**Reply by ekwoka** [reply]
running initTree directly is totally fine.

It's the same thing Alpine does.

**Reply by nikitaourazbaev** [reply]
Thanks!
